File: app/src/main/java/net/daverix/ajvm/jvm/IndexedILoadOperation.java
package net.daverix.ajvm.jvm;


import net.daverix.ajvm.Frame;
import net.daverix.ajvm.ByteCodeReader;

import java.io.IOException;

public class IndexedILoadOperation implements ByteCodeOperation {
    private final int index;

    public IndexedILoadOperation(int index) {
        this.index = index;
    }

    @Override
    public void execute(ByteCodeReader reader,
                        int indexOfBytecode,
                        Frame currentFrame) throws IOException {
        Object variable = currentFrame.getLocalVariable(index);
        if (!(variable instanceof Integer))
            throw new IllegalStateException("variable " + variable + " is not an Integer");

        currentFrame.push(variable);
    }
}

File: app/src/main/java/net/daverix/ajvm/operation/IndexedILoadOperation.kt
package net.daverix.ajvm.operation


import net.daverix.ajvm.Frame
import net.daverix.ajvm.ByteCodeReader

import java.io.IOException

class IndexedILoadOperation(private val index: Int) : ByteCodeOperation {

    @Throws(IOException::class)
    override fun execute(reader: ByteCodeReader,
                         indexOfBytecode: Int,
                         currentFrame: Frame) {
        val variable = currentFrame.getLocalVariable(index)
        if (variable !is Int)
            throw IllegalStateException("variable $variable is not an Integer")

        currentFrame.push(variable)
    }
}
